File: frontend/src/types/bigdata.ts
/** Big data module types */

export type JobType = 'mysql_to_hdfs' | 'build_marts' | 'batch_predict' | 'custom_hive';

export type JobStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled';

export interface Job {
  job_id: string;
  job_type: JobType;
  status: JobStatus;
  started_at: string | null;
  finished_at: string | null;
  params?: Record<string, unknown> | null;
  duration_ms?: number | null;
  rows_processed?: number | null;
  submitted_by?: number | null;
}

export interface JobDetail extends Job {
  log_tail?: string[];
  logs?: string;
}

export interface JobListQuery {
  page?: number;
  page_size?: number;
}

export interface JobSubmitRequest {
  job_type: JobType;
  params?: Record<string, unknown>;
}

export interface JobSubmitResponse {
  job_id: string;
  status: JobStatus;
}

export interface HdfsEntry {
  name: string;
  type: 'file' | 'directory';
  size: number;
  modified: string;
}

export interface HdfsListResponse {
  entries: HdfsEntry[];
}

export interface HdfsPreviewResponse {
  content: string;
}

export interface HiveQueryRequest {
  sql: string;
  limit?: number;
}

export interface HiveQueryResponse {
  columns: string[];
  rows: unknown[][];
}

/** Feature dictionary used for ML inference */
export interface FeatureDict {
  AGE: number;
  IS_FEMALE: number;
  RACE: number;
  SCHLYRS: number;
  SELF_HEALTH: number;
  HEALTH_CHANGE: number;
  FALL_2YR: number;
  PAIN: number;
  BMI_CATEGORY: number;
  MEMORY_RATING: number;
  MEMORY_CHANGE: number;
  SERIAL7_SCORE: number;
  DATE_NAMING: number;
  ADL_SCORE: number;
  HOSPITAL_STAY: number;
  NURSING_HOME: number;
  HOME_HEALTH: number;
  HAS_USUAL_CARE: number;
  NUM_HOSPITAL_STAYS: number;
  DOCTOR_VISITS: number;
}

export interface Prediction {
  high_risk_prob: number;
  high_risk: boolean;
  followup_prob: number;
  followup_needed: boolean;
  health_score: number;
}

/** Historical prediction record entry */
export interface PredictionRecord {
  id: number;
  elder_id: number;
  predicted_at: string;
  high_risk_prob: number;
  followup_prob: number;
  high_risk: boolean;
  followup_needed: boolean;
  health_score: number;
}

/** Feature contribution breakdown returned with a prediction. */
export interface FeatureContribution {
  key: string;
  label: string;
  value: number;
  z_score: number;
  direction: 'higher' | 'lower';
}

/** Predict + attribution response. */
export interface PredictionWithContributions extends Prediction {
  contributions?: FeatureContribution[];
}

/** Autofill payload for an elder before running inference. */
export interface MLFeaturePayload {
  elder_id: number;
  features: Record<string, number | null>;
  sources: Record<string, 'elder' | 'health_record' | 'survey' | null>;
  missing: string[];
}

/** Response of POST /ml/predict_by_elder. */
export interface PredictByElderResponse extends PredictionWithContributions {
  used_features: Record<string, number>;
  sources: Record<string, 'elder' | 'health_record' | 'survey' | null>;
  missing_before_overrides: string[];
}

export interface HiveSavedQuery {
  id: number;
  user_id: number;
  name: string;
  sql: string;
  description?: string | null;
  created_at: string;
  updated_at: string;
}

export interface HiveQueryHistoryEntry {
  id: number;
  user_id: number;
  sql: string;
  row_count: number;
  duration_ms: number;
  status: 'success' | 'failed';
  error_message?: string | null;
  created_at: string;
}

export interface AnalyticsRiskDistribution {
  items: { key: string; label: string; count: number }[];
  total: number;
}

export interface AnalyticsFollowupCompletion {
  items: { date: string; todo: number; in_progress: number; completed: number }[];
  days: number;
}

export interface AnalyticsRegionalBreakdown {
  items: { region: string; count: number }[];
}

export interface AnalyticsPipelineHealth {
  items: {
    stage: string;
    status: string;
    job_id: string | null;
    duration_ms?: number | null;
    rows_processed?: number | null;
    finished_at?: string | null;
  }[];
}

export interface AnalyticsPredictionTrend {
  items: {
    date: string;
    avg_health_score: number;
    high_risk_count: number;
    total: number;
  }[];
  days: number;
}

File: backend/app/schemas/bigdata.py
"""Pydantic schemas for big data, Spark, Hive, HDFS, and ML inference endpoints."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


# ---- Jobs ----

class BigDataJobCreate(BaseModel):
    """Request body to submit a big data job."""

    job_type: str = Field(
        ...,
        description="One of: mysql_to_hdfs, build_marts, batch_predict, custom_hive",
    )
    params: Optional[dict] = None


class BigDataJobResponse(BaseModel):
    """Big data job response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    job_id: str
    job_type: str
    status: str
    params: Optional[dict] = None
    log_path: str = ""
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    submitted_by: Optional[int] = None
    duration_ms: Optional[int] = None
    rows_processed: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class BigDataJobDetail(BigDataJobResponse):
    """Job detail with recent log lines."""

    log_tail: list[str] = Field(default_factory=list)


# ---- HDFS ----

class HdfsFileEntry(BaseModel):
    """A single entry returned from WebHDFS LISTSTATUS."""

    name: str
    type: str
    size: int = 0
    modified: Optional[int] = None
    owner: str = ""
    group: str = ""
    permission: str = ""


class HdfsListResponse(BaseModel):
    """HDFS directory listing response."""

    path: str
    entries: list[HdfsFileEntry]


class HdfsPreviewResponse(BaseModel):
    """HDFS file head preview."""

    path: str
    lines: list[str]
    truncated: bool = False


# ---- Hive ----

class HiveQueryRequest(BaseModel):
    """Hive SELECT query request."""

    sql: str
    limit: int = Field(200, ge=1, le=10000)


class HiveQueryResponse(BaseModel):
    """Hive query result."""

    columns: list[str]
    rows: list[list[Any]]


# ---- ML inference ----

class MLPredictRequest(BaseModel):
    """Single-record ML inference request. Unknown keys default to 0."""

    model_config = ConfigDict(extra="allow")

    features: dict[str, float] = Field(default_factory=dict)


class FeatureContribution(BaseModel):
    """Top-k attribution entry for a single feature."""

    key: str
    label: str
    value: float
    z_score: float
    direction: str  # "higher" | "lower"


class MLPredictResponse(BaseModel):
    """Single-record ML inference result."""

    high_risk_prob: float
    high_risk: bool
    followup_prob: float
    followup_needed: bool
    health_score: float
    contributions: Optional[list[FeatureContribution]] = None


class MLFeaturePayload(BaseModel):
    """Autofilled feature dict for an elder + gap list."""

    elder_id: int
    features: dict[str, Optional[float]] = Field(default_factory=dict)
    sources: dict[str, Optional[str]] = Field(default_factory=dict)
    missing: list[str] = Field(default_factory=list)


class HiveSavedQueryCreate(BaseModel):
    name: str = Field(..., max_length=128)
    sql: str
    description: Optional[str] = None


class HiveSavedQueryUpdate(BaseModel):
    name: Optional[str] = None
    sql: Optional[str] = None
    description: Optional[str] = None


class HiveSavedQueryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    name: str
    sql: str
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class HiveQueryHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    sql: str
    row_count: int
    duration_ms: int
    status: str
    error_message: Optional[str] = None
    created_at: datetime


class MLBatchPredictRequest(BaseModel):
    """Batch ML inference request."""

    records: list[dict[str, float]]


class PredictionResultResponse(BaseModel):
    """Persisted prediction result row."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    elder_id: int
    high_risk_prob: float
    high_risk: bool
    followup_prob: float
    followup_needed: bool
    health_score: float
    predicted_at: datetime
    created_at: datetime
    updated_at: datetime


# ---- Pipeline freshness (business-facing wrapper over pipeline-health) ----


class StageFreshness(BaseModel):
    """Business-facing freshness summary for a single pipeline stage."""

    stage: str
    display_name: str
    description: str
    status: str
    job_id: Optional[str] = None
    finished_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    rows_processed: Optional[int] = None
    freshness_seconds: Optional[int] = None
    freshness_label: str
    freshness_tone: str  # "fresh" | "aging" | "stale" | "never" | "running"


class PipelineSchedule(BaseModel):
    """Declarative config of the daily auto-run. Source of truth is UTC."""

    enabled: bool
    utc_time: str  # "HH:MM" in UTC
    source: str  # "db" | "env"
    next_run_at: Optional[datetime] = None  # UTC


class PipelineScheduleUpdate(BaseModel):
    """Admin PUT body. utc_time is HH:MM in UTC (24h)."""

    enabled: bool
    utc_time: str


class PipelineFreshnessResponse(BaseModel):
    """Aggregated freshness for the three-stage pipeline."""

    stages: list[StageFreshness]
    has_running_pipeline: bool
    pipeline_run_id: Optional[str] = None
    running_stage: Optional[str] = None
    schedule: Optional[PipelineSchedule] = None


class PipelineRunResponse(BaseModel):
    """Result of submitting a one-click pipeline run."""

    pipeline_run_id: str
    job_ids: list[str]
    reused: bool  # True when an existing running pipeline was returned
